### ISSUE TITLE
Too many implementations of state tomography

### ISSUE THREAD

**Issue by viathor** [quantumlib/Cirq#2737]
We have three implementations of state tomography:
- for 1-qubit states,
- for 2-qubit states,
- for n-qubit states for arbitrary n.

The last one isn't well tested (yet). Once it is, we should compare performance and if there isn't a clear advantage to the fixed-size implementations we should remove them.

**Comment by mpharrigan** [comment]
Should we keep this open? The linked issue fixes the bug in n-qubit tomography but the larger issue is still present

**Comment by viathor** [comment]
Yes, thank you. My comment on PR #2739 incorrectly said that it fixed #2737. It should have said that it fixed #2738.

**Comment by dabacon** [comment]
There can never be too many implementations of state tomography :)

But yes we should unify these standard method of state tomography (there are many others).

**Comment by MichaelBroughton** [comment]
Is this still an open concern ?

**Comment by maffoo** [comment]
I'd say yes this is still a concern. It'd be nice to unify these tomography functions, but not a blocker for 1.0.

**Comment by NoureldinYosri** [comment]
there is a need to consolidate the different implemtations from from a code health point (3 diffferent implementations) as well as performance and numerical stability points. 

@cyrikroivel would you like to add your optimized implementation to Cirq? we can later consolidate the implementations around it.

**Comment by CyrikRoivel** [comment]
Yes it could be useful to add it to cirq. Also, I think that it could be valuable to have multiple implementations depending on the desired use case. There is an existing state_tomography that is sampling the complete probability distribution of strings output by a circuit and then doing a max likelihood estimate to determine the density matrix. This could be desirable for smaller systems, maybe up to 6 qubits or so, but one starts to be limited by memory as the approach necessitates some redundant information. Because I wanted to reach a higher number of qubits, I implement a simpler version that measures something closer to the minimal number of observables required to construct the density matrix, so no max likelihood is done and fewer numbers need to be stored. I am still working on some tests though so I'd like to hold off from putting it on cirq for a week or so.

**Comment by NoureldinYosri** [comment]
@CyrikRoivel would you like to take over this task ... we can add extra arguments for these different variations

**Comment by arettig** [comment]
From what I can see, we have [1-qubit](https://github.com/quantumlib/Cirq/blob/daf1fa685996deb3874d7fb54c760ff01b0ba57f/cirq-core/cirq/experiments/qubit_characterizations.py#L642), [2-qubit](https://github.com/quantumlib/Cirq/blob/daf1fa685996deb3874d7fb54c760ff01b0ba57f/cirq-core/cirq/experiments/qubit_characterizations.py#L691), and [n-qubit](https://github.com/quantumlib/Cirq/blob/daf1fa685996deb3874d7fb54c760ff01b0ba57f/cirq-core/cirq/experiments/n_qubit_tomography.py#L137) state tomography implementations. Ideally, we could get rid of the specialized 1 and 2 qubit implementations provided the n-qubit code has comparable speed for these cases.

I ran some quick benchmarks using some very simple test cases:

| Implementation | Median Time (IQR) |
| :--- | :---: | 
| 1-Qubit | 19.4 (0.1) s | 
| n-Qubit (n=1) | 37.1 (0.2) s | 
||||
| 2-Qubit | 16.4 (0.1) s | 
| n-Qubit (n=2) | 16.3 (0.0) s| 
||||
| n-Qubit (n=5) | 15.2 (0.2) s| 
 


It looks like the n-qubit state tomography is slower than the specialized 1-qubit state tomography, but it is comparable in speed to the 2-qubit implementation. Interestingly, taking a look at the profiler shows that this is because the 2-qubit and n-qubit implementations use the `results.histogram()` method, which is slow due to calling `fold_func()` for every sample in a python loop. Of course, with a nontrivial state preparation circuit this post-processing time would probably be negligible anyways.

So all that to say that in most cases the timing difference is probably negligible, but there are some cases where the state preparation circuit is tiny and `repetitions` is large where the 1-qubit state tomography will be noticeably faster than the n-qubit general implementation.

I tested using a vectorized version of `big_endian_bits_to_int` which is the default `fold_func` and that got rid of the `histogram()` bottleneck. So I'll add that and then run everything through the n-qubit state tomography.

<details>

<summary> Test Cases </summary>

Run benchmark via:
```
pytest -p no:randomly --enable-slow-tests  \
    --benchmark-disable-gc --benchmark-warmup=yes \
    benchmarks/tomography_perf.py --benchmark-enable
```


tomography_perf.py:
```
import pytest

import cirq
from cirq.experiments.qubit_characterizations import (
    single_qubit_state_tomography,
    two_qubit_state_tomography
)
from cirq.experiments.n_qubit_tomography import state_tomography


class TestStateTomography:
    """State Tomography Benchmarks."""

    group = "tomography"

    @pytest.mark.parametrize("repetitions", [6000000])
    @pytest.mark.benchmark(group=group)
    def test_one_qubit_specialized(self, benchmark, repetitions):
        """Benchmark state tomography using specialized one qubit implementation."""
        q0 = cirq.LineQubit(0)
        c1 = cirq.Circuit([cirq.H(q0)])
        sampler = cirq.Simulator()
        f = lambda: single_qubit_state_tomography(sampler, q0, c1, repetitions)
        benchmark(f)

    @pytest.mark.parametrize("repetitions", [6000000])
    @pytest.mark.benchmark(group=group)
    def test_one_qubit_general(self, benchmark, repetitions):
        """Benchmark state tomography using general one qubit implementation."""
        q0 = cirq.LineQubit(0)
        c1 = cirq.Circuit([cirq.H(q0)])
        sampler = cirq.Simulator()
        f = lambda: state_tomography(sampler, [q0], c1, repetitions)
        benchmark(f)

    @pytest.mark.parametrize("repetitions", [600000])
    @pytest.mark.benchmark(group=group)
    def test_two_qubit_specialized(self, benchmark, repetitions):
        """Benchmark state tomography using specialized two qubit implementation."""
        qs = cirq.LineQubit.range(2)
        c2 = cirq.Circuit([cirq.H(qs[0]), cirq.H(qs[1])])
        sampler = cirq.Simulator()
        f = lambda: two_qubit_state_tomography(sampler, qs[0], qs[1], c2, repetitions)
        benchmark(f)

    @pytest.mark.parametrize("repetitions", [600000])
    @pytest.mark.benchmark(group=group)
    def test_two_qubit_general(self, benchmark, repetitions):
        """Benchmark state tomography using general two qubit implementation."""
        qs = cirq.LineQubit.range(2)
        c2 = cirq.Circuit([cirq.H(qs[0]), cirq.H(qs[1])])
        sampler = cirq.Simulator()
        f = lambda: state_tomography(sampler, qs, c2, repetitions)
        benchmark(f)

    @pytest.mark.parametrize("repetitions", [10000])
    @pytest.mark.benchmark(group=group)
    def test_n_qubit_general(self, benchmark, repetitions):
        """Benchmark state tomography using general two qubit implementation."""
        qs = cirq.LineQubit.range(5)
        c = cirq.Circuit([cirq.H(q) for q in qs])
        sampler = cirq.Simulator()
        f = lambda: state_tomography(sampler, qs, c, repetitions)
        benchmark(f)

```

</details>

<details>

<summary> 1-qubit Profiler comparison </summary>

| n-qubit (n=1) | 1-qubit |
| ------------- | ------------- |
|<img alt="Image" src="https://github.com/user-attachments/assets/5c93a7e6-5ff2-468a-9cf7-ba77ff71b2c7" /> | <img  alt="Image" src="https://github.com/user-attachments/assets/6903f197-8f83-4091-8865-fc9a022f6469" />|

</details>

<details>

<summary> 2-qubit Profiler comparison </summary>

| n-qubit (n=2) | 2-qubit |
| ------------- | ------------- |
|  <img alt="Image" src="https://github.com/user-attachments/assets/f1a1e0ed-f7a9-4600-bf91-d4a117b2f030" />=|  <img alt="Image" src="https://github.com/user-attachments/assets/25abd024-d795-4076-a1a4-c95ef9cfb889" />|

</details>
